packages/use-dataloader/src/useDataLoader.js

```diff
@@ -0,0 +1,164 @@
+import {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useReducer,
+  useRef,
+} from 'react'
+import { useDataLoaderContext } from './DataLoaderProvider'
+import { ActionEnum, StatusEnum } from './constants'
+import reducer from './reducer'
+
+const Actions = {
+  createReset: ({ data }) => ({ type: ActionEnum.RESET, data }),
+  createOnLoading: () => ({ type: ActionEnum.ON_LOADING }),
+  createOnSuccess: data => ({ type: ActionEnum.ON_SUCCESS, data }),
+  createOnUpdateData: data => ({ type: ActionEnum.ON_UPDATE_DATA, data }),
+  createOnError: error => ({ type: ActionEnum.ON_ERROR, error }),
+}
+
+const useDataLoader = (
+  key,
+  method,
+  {
+    onSuccess,
+    onError,
+    initialData,
+    pollingInterval,
+    enabled = true,
+    reloadOnKeyChange = true,
+    keepOldData = true,
+  },
+) => {
+  const {
+    addCachedData,
+    addReload,
+    clearReload,
+    getCachedData,
+  } = useDataLoaderContext()
+  const [{ status, data, error }, dispatch] = useReducer(reducer, {
+    status: StatusEnum.IDLE,
+    data: initialData,
+    error: undefined,
+  })
+
+  const oldDataRef = useRef()
+  const keyRef = useRef()
+  const methodRef = useRef(method)
+  const onSuccessRef = useRef(onSuccess)
+  const onErrorRef = useRef(onError)
+
+  const isLoading = useMemo(() => status === StatusEnum.LOADING, [status])
+  const isIdle = useMemo(() => status === StatusEnum.IDLE, [status])
+  const isSuccess = useMemo(() => status === StatusEnum.SUCCESS, [status])
+  const isError = useMemo(() => status === StatusEnum.ERROR, [status])
+  const isPolling = useMemo(
+    () => enabled && pollingInterval && (isSuccess || isLoading),
+    [isSuccess, isLoading, enabled, pollingInterval],
+  )
+
+  const updateData = useCallback(
```

Comment:
udpateDate should not be used outside of the dataLoader, data is always true from api and not from our side if we can manipulate it